larq_compute_engine/core/types.h

```diff
@@ -9,6 +9,21 @@
 namespace compute_engine {
 namespace core {
 
+#if defined(__GNUC__)
+#define LCE_LIKELY(condition) (__builtin_expect(!!(condition), 1))
+#define LCE_UNLIKELY(condition) (__builtin_expect(condition, 0))
+#else
+#define LCE_LIKELY(condition) (condition)
+#define LCE_UNLIKELY(condition) (condition)
+#endif
+
+// Check that 0 <= index < limit using a single comparison, assuming
+// that 0 <= limit if Index is signed.  Intended for use in performance
+// critical contexts where 0 <= index < limit is almost always true.
+inline bool FastBoundsCheck(const int index, const int limit) {
+  return LCE_LIKELY((unsigned)index < (unsigned)limit);
+}
```

Comment:
I think this might rely on undefined behaviour. Definitely prior to [C++20](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0907r4.html), there was no requirement for signed integer types to use twos complement, which I think this trick relies on?

If this code were being used in performance-critical code I think there would be a strong argument for it, but the two uses are in `prepare.h` in functions which only get called once at the very start, not per-inference.

Reply:
Makes sense, I remove their usage.